server/src/pull/cvr.ts

```diff
@@ -0,0 +1,229 @@
+  executor: Executor,
+  table: keyof typeof TableOrdinal,
+  clientGroupID: string,
+  order: number,
+  limit: number,
+) {
+  //   sql = /*sql*/ `SELECT * FROM "${table}" t WHERE NOT EXISTS (
+  //       SELECT 1 FROM "cvr_entry" WHERE "cvr_entry"."row_id" = t."id" AND
+  //       "cvr_entry"."row_version" = t."rowversion" AND
+  //       "cvr_entry"."client_group_id" = $1 AND
+  //       "cvr_entry"."order" <= $2 AND
+  //       "cvr_entry"."tbl" = $3
+  //     ) LIMIT $4`;
+  // The below query runs in ~40ms for issues vs the above takes 16 seconds for issues.
+  const sql = /*sql*/ `SELECT *
+      FROM "${table}" t
+      WHERE (t."id", t."rowversion") NOT IN (
+        SELECT "row_id", "row_version"
+        FROM "cvr_entry"
+        WHERE "client_group_id" = $1
+          AND "order" <= $2
+          AND "tbl" = $3
+      )
+      LIMIT $4;`;
+
+  const params = [clientGroupID, order, TableOrdinal[table], limit];
+  return executor(sql, params);
+}
+
+export const TableOrdinal = {
+  issue: 1,
+  description: 2,
+  comment: 3,
+} as const;
+
+export function findDeletions(
+  executor: Executor,
+  table: keyof typeof TableOrdinal,
+  clientGroupID: string,
+  order: number,
+  limit: number,
+) {
+  // Find rows that are in the CVR table but not in the actual table.
+  // Exclude rows that have a delete entry already.
+  // TODO: Maybe we can remove the second `NOT EXISTS` if we prune the CVR table.
+  // This pruning would mean that we need to record the delete against the
+  // current CVR rather than next CVR. If a request comes in for that prior CVR,
+  // we return the stored delete records and do not compute deletes.
+  return executor(
+    /*sql*/ `SELECT "row_id" FROM "cvr_entry"
+    WHERE "cvr_entry"."tbl" = $1 AND NOT EXISTS (
+      SELECT 1 FROM "${table}" WHERE id = "cvr_entry"."row_id"
+    ) AND
+    "cvr_entry"."client_group_id" = $2 AND
+    "cvr_entry"."order" <= $3 
+    AND NOT EXISTS (
+      SELECT 1 FROM "cvr_delete_entry" WHERE "cvr_delete_entry"."tbl" = $1 AND "cvr_delete_entry"."row_id" = "cvr_entry"."row_id"
+      AND "cvr_delete_entry"."client_group_id" = $2 AND "cvr_delete_entry"."order" <= $3
+    ) LIMIT $4`,
+    [TableOrdinal[table], clientGroupID, order, limit],
+  );
+}
+
+export async function dropCVREntries(
+  executor: Executor,
+  clientGroupID: string,
+  order: number,
+) {
+  await Promise.all([
+    executor(
+      /*sql*/ `DELETE FROM "cvr_entry" WHERE "client_group_id" = $1 AND "order" > $2`,
+      [clientGroupID, order],
+    ),
+    executor(
+      /*sql*/ `DELETE FROM "cvr_delete_entry" WHERE "client_group_id" = $1 AND "order" > $2`,
+      [clientGroupID, order],
+    ),
+  ]);
+}
+
+export async function recordUpdates(
+  executor: Executor,
+  table: keyof typeof TableOrdinal,
+  clientGroupID: string,
+  order: number,
+  updates: SearchResult[],
+) {
+  if (updates.length === 0) {
+    return;
+  }
+
+  const placeholders = [];
+  const values = [];
+  const stride = 5;
```

Comment:
Why build up these two huge arrays. Can you instead allocate a single array of length five for the values and use it over and over? There's no need for the placeholder array in the first place -- you can just hardcode the placeholders in the SQL string.

Reply:
The intent was to reduce the number of calls to the DB.

E.g., if we're recording a batch of 3,000 things then it'll just issue 1 query with all 3k values.

Not building up these arrays would mean sending 3,000 separate insert queries.

E.g., changing from this:

```ts
for (let i = 0; i < updates.length; i++) {
    placeholders.push(
      `($${i * stride + 1}, $${i * stride + 2}, $${i * stride + 3}, $${
        i * stride + 4
      }, $${i * stride + 5})`,
    );
    values.push(
      clientGroupID,
      order,
      TableOrdinal[table],
      updates[i].id,
      updates[i].version,
    );
  }

  await executor(
    /*sql*/ `INSERT INTO client_view_entry (
    "client_group_id",
    "client_view_version",
    "entity",
    "entity_id",
    "entity_version"
  ) VALUES ${placeholders.join(
    ', ',
  )} ON CONFLICT ("client_group_id", "entity", "entity_id") DO UPDATE SET
    "entity_version" = excluded."entity_version",
    "client_view_version" = excluded."client_view_version"
  `,
    values,
  );
```

to this:

```ts
await Promise.all(updates.map(u => {
    await executor(
      /*sql*/ `INSERT INTO client_view_entry (
      "client_group_id",
      "client_view_version",
      "entity",
      "entity_id",
      "entity_version"
    ) VALUES (?, ?, ?, ?, ?) ON CONFLICT ("client_group_id", "entity", "entity_id") DO UPDATE SET
      "entity_version" = excluded."entity_version",
      "client_view_version" = excluded."client_view_version"
    `,
     [clientGroupID, order, table, u.id, u.version],
    );
}))
```

So it's either:
1. Create and await 3k promises (/ some chunking of this)
2. Or create a large string of 3k entries and 1 promise.

Reply:
I seeeeee. I totally missed this, thank you for explaining.

Is it actually faster? Did you measure it? If significantly faster, I support this complexity but if it's marginal I'd prefer the more direct `Promise.all()` impl.

Reply:
Nevermind -- https://github.com/vitaly-t/pg-promise/wiki/Performance-Boost confirms that it's likely to be much slower (and also has some helpers that may or may not be useful to you).